advtraj/integrate/forward.py

```diff
@@ -1,25 +1,575 @@
+        options.update(pioptions)
+
+    # defaults
+    maxiter = options.get("maxiter", 500)
+    miniter = options.get("miniter", 10)
+    relax = options.get("relax", 0.8)
+    relax_reduce = options.get("relax_reduce", 0.9)
+    tol = options.get("tol", 0.01)
+    opt_one_step = options.get("opt_one_step", False)
+    disp = options.get("disp", False)
+    norm = options.get("norm", "max_abs_error")
+    # = options.get('', )
+
+    grid_spacing = np.array(
+        [
+            ds_position_scalars["x"].attrs["dx"],
+            ds_position_scalars["y"].attrs["dy"],
+            ds_position_scalars["z"].attrs["dz"],
+        ]
+    )
+
+    grid_spacing = grid_spacing[:, np.newaxis]
+
+    def get_error_norm(
+        scalars, pos_org, pos_est, interpolator=interpolator, interp_order=interp_order
+    ):
+        dist = _calc_backtrack_origin_dist(
+            scalars,
+            pos_org,
+            pos_est,
+            interpolator=interpolator,
+            interp_order=interp_order,
+        )
+
+        ndist = dist / grid_spacing
+
+        if norm == "max_abs_error":
+            err = np.linalg.norm(ndist.flatten(), ord=np.inf)
+        if norm == "mean_abs_error":
+            err = np.mean(np.abs(ndist))
+        else:
+            err = np.linalg.norm(ndist.flatten()) / np.sqrt(ndist.size)
+
+        return dist, ndist, err
+
+    # First find error in first guess.
+    dist, ndist, err = get_error_norm(
+        ds_position_scalars,
+        ds_traj_posn_org,
+        ds_traj_posn_first_guess,
+        interpolator=interpolator,
+        interp_order=interp_order,
+    )
+
+    ds_traj_posn_next = ds_traj_posn_first_guess
+
+    # Now setup iteration
+    niter = 0
+    not_converged = True
+    while not_converged:
+
+        # We have found convergence is faster if we move slightly less
+        # than the residual distance; this is the relax parameter,
+        # generally <1.
+        # Restrict change to no more than 1 grid box.
+        delta = np.clip(dist * relax, -grid_spacing, grid_spacing)
+
+        for i, c in enumerate("xyz"):
+            if delta.shape[1] == 1:
+                ds_traj_posn_next[c] += np.squeeze(delta[i])
+            else:
+                ds_traj_posn_next[c] += delta[i, ...]
+
+        dist, ndist, err = get_error_norm(
+            ds_position_scalars,
+            ds_traj_posn_org,
+            ds_traj_posn_next,
+            interpolator=interpolator,
+            interp_order=interp_order,
+        )
+
+        niter += 1
+        not_converged = err > tol
+
+        if niter >= maxiter:
+            break
+        # If we need a lot of iterations, reduce the relaxation factor.
+        if niter % miniter == 0:
+            relax *= relax_reduce
+
+    # Are we converged?
+    if niter >= maxiter:
+        # Select out those trajectories that have not converged.
+        ncm = np.abs(ndist) > tol
+        not_conv_mask = ncm[0, :] | ncm[1, :] | ncm[2, :]
+        print(
```

Comment:
I think we should use a global logger (like `loguru` I mentioned above) and then call `logger.info(...)` here instead. Would remove an argument to these functions

Reply:
I very much agree with using a logger (and thanks for pointing me at `loguru `- I've used `logging`, the bog standard.

However, van we postpone to another pull as I think we need to discuss what we want to log to where. I'll have a play with loguru.